File: Heap/Easy/min_cost_of_ropes.py
# https://practice.geeksforgeeks.org/problems/minimum-cost-of-ropes-1587115620/1#
#User function Template for python3
from heapq import heappush, heappop, heapify

class Solution:
    #Function to return the minimum cost of connecting the ropes.
    def minCost(self,arr,n) :
        if(n == 1):
            return 0
        
        heapify(arr)
                    
        ans = 0
        while(len(arr) > 1):
            top = heappop(arr)
            top += heappop(arr)
            ans += top
            heappush(arr, top)
            
        return ans

#{ 
#  Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys
import heapq
from collections import  defaultdict
# Contributed by : Nagendra Jha

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER

@atexit.register

def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())

if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        n = int(input())
        a = list(map(int, input().strip().split()))
        ob=Solution()
        print(ob.minCost(a,n))
# } Driver Code Ends

File: Heap/Easy/min_cost_of_ropes.cpp
// https://practice.geeksforgeeks.org/problems/minimum-cost-of-ropes-1587115620/1#
// { Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


 // } Driver Code Ends
class Solution
{
    public:
    //Function to return the minimum cost of connecting the ropes.
    long long minCost(long long arr[], long long n) {
        if(n == 1) return 0;
        
        long long ans = 0;
        priority_queue<long long, vector<long long>, greater<long long>> pq;
        
        for(int i = 0; i < n; i++) pq.push(arr[i]);
        
        while(pq.size() > 1)
        {
            long long top = pq.top();
            pq.pop();
            top += pq.top();
            pq.pop();
            
            ans += top;
            pq.push(top);
        }
        
        return ans;
    }
};


// { Driver Code Starts.

int main() {
    long long t;
    cin >> t;
    while (t--) {
        long long n;
        cin >> n;
        long long i, a[n];
        for (i = 0; i < n; i++) {
            cin >> a[i];
        }
        Solution ob;
        cout << ob.minCost(a, n) << endl;
    }
    return 0;
}
  // } Driver Code Ends